compiler/rustc_trait_selection/src/solve/fulfill/derive_errors.rs

```diff
@@ -506,6 +506,25 @@ impl<'tcx> ProofTreeVisitor<'tcx> for BestObligation<'tcx> {
             }
         }
 
+        // HACK: When a higher-ranked projection goal fails, check that the corresponding
+        // higher-ranked trait goal holds or not. This is because the process of instantiating
+        // and then re-canonicalizing the binder of the projection goal forces us to be unable
+        // to see that the leak check failed in the nested `NormalizesTo` goal, so we don't
+        // fall back to the regular machinery that should catch when a projection goal fails
+        // due to an unsatisfied trait goal.
```

Comment:
the regular machinery is just the `alias_is_rigid`/`proven_via` check inside of the solver itself?

r=me, if u think adding that to the comment is useful, please do

Reply:
Basically yea, will add comment